pxtlib/sprite-editor/spriteEditor.ts

```diff
@@ -313,62 +314,107 @@ namespace pxtsprite {
             this.closeHandler = handler;
         }
 
-        switchIconBack(tool: PaintTool) {
-            let btn = (this.sidebar.getButtonForTool(tool) as TextButton);
-            if (tool == PaintTool.Rectangle) {
-                //Change icon back to square
-                btn.setText("\uf096");
-                btn.title(lf("Rectangle"));
-            } else if (tool == PaintTool.Normal) {
-                //Change icon back to pencil
-                btn.setText("\uf040");
-                btn.title(lf("Pencil"));
+        switchIconTo(tool: PaintTool) {
+            if (this.activeTool === tool) return;
+
+            const btn = (this.sidebar.getButtonForTool(tool) as TextButton);
+
+            switch (tool) {
+                case PaintTool.Rectangle:
+                    updateIcon(btn, "\uf096", lf("Rectangle"));
+                    break;
+                case PaintTool.Circle:
+                    updateIcon(btn, "\uf10c", lf("Circle"));
+                    break;
+                case PaintTool.Normal:
+                    updateIcon(btn, "\uf040", lf("Pencil"));
+                    break;
+                case PaintTool.Line:
+                    updateIcon(btn, "\uf07e", lf("Line"));
+                    break;
+                default:  // no alternate icon, do not change
+                    return;
             }
-            btn.onClick(() => this.sidebar.setTool(tool));
-            if ((this.activeTool == PaintTool.Circle && tool == PaintTool.Rectangle)
-                    || (this.activeTool == PaintTool.Line && tool == PaintTool.Normal)) {
-                this.setActiveTool(tool);
+
+            btn.onClick(() => {
+                if (tool != PaintTool.Circle && tool != PaintTool.Line) {
+                    this.setIconsToDefault();
+                    this.sidebar.setTool(tool);
+                }
+            });
+
+            function updateIcon(button: TextButton, text: string, title: string) {
+                const shortcut = getPaintToolShortcut(tool);
+
+                button.setText(text);
+                button.title(title);
+                button.shortcut(shortcut);
             }
         }
 
+        setIconsToDefault() {
+            this.switchIconTo(PaintTool.Rectangle);
+            this.switchIconTo(PaintTool.Normal);
+        }
+
         private keyDown = (event: KeyboardEvent) => {
+            if (event.key === "Undo" || (event.ctrlKey && event.key === "z")) {
```

Comment:
This was already in an event handler, but doesn't actually run -  the blockly undo is run and closes this before the event would be fired. Anyone happen to know of a way to prevent that behavior (preventDefault doesn't.)

Reply:
Nevermind, fixed by moving the handling of the event to the capture phase and stopping propagation of the event.